### PR TITLE
Add digital dynamical decoupling high-level functions

### LINKED ISSUE #1223: Add `execute_with_ddd()`, `mitigate_executor()`, and `ddd_decorator()` for dynamical decoupling
Issue Description
-----------------

Add `execute_with_ddd()`, `mitigate_executor()`, and `ddd_decorator()` to `ddd.py`

Additional References
---------------------

See the relevant RFC https://docs.google.com/document/d/1cRwFCTn6kUjI1P0kNydtevxIYtE4r8Omd_iWK0Pe8qo/edit#

### PR DESCRIPTION
**Pull request by andreamari**
Description
-----------
This PR adds digital dynamical decoupling high-level functions.
Fixes #1223 

Checklist
---------

Check off the following once complete (or if not applicable) after opening the PR. The PR will be reviewed once this checklist is complete and all tests are passing.

- [x] I added unit tests for new code.
- [x] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [x] I used [Google-style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) docstrings for functions.
- [x] I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.
- [x] Added myself / the copyright holder to the AUTHORS file

If some items remain, you can mark this a [draft pull request](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

License
-------

- [x] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.

### CODE REVIEW

**Comment by Aaron-Robertson** [comment]
So far looks very nice @andreamari! One note other than the unused argument above would be to add some testing around the `rule_args` optional argument.

Docs changes look good/are clearer overall, but would it be better to add them separately? Only concern is it could be confusing in history (though unlikely that the history around those changes will matter, so I don't have strong feelings either way)

**Comment by andreamari** [comment]
> So far looks very nice @andreamari! One note other than the unused argument above would be to add some testing around the `rule_args` option.


Hi @Aaron-Robertson, I added a new test for `rule_args`. I added it to the decorator, so it percolates down to `execute_with_ddd` and in this way we are more sure that the option works with all the three new functions.

> Docs changes look good/are clearer overall, but would it be better to add them separately? Only concern is it could be confusing in history (though unlikely that the history around those changes will matter, so I don't have strong feelings either way)

You are right, what I did is bad practice. However changes in docstrings are minimal and we have limited time, **only for this time** :smile: , I would prefer to save time and avoid making  2 new PRs.

File: mitiq/ddd/ddd.py
Comment (andreamari) [reply mitiq/ddd/ddd.py:42]: Good catch!

**Review by Aaron-Robertson — APPROVED** [review]

**Comment by Misty-W** [comment]
🎉

**Comment by andreamari** [comment]
Hi @Misty-W, I know that you are waiting for this PR for making the release. 
I need to go now for a few hours, but with 99% probability tests will eventually be all green.
So, after tests are completed, feel free to merge such that you can go on with the rest.

**Comment by Misty-W** [comment]
> Hi @Misty-W, I know that you are waiting for this PR for making the release. I need to go now for a few hours, but with 99% probability tests will eventually be all green. So, after tests are completed, feel free to merge such that you can go on with the rest.

will do, thanks @andreamari !

**Comment by andreamari** [comment]
Let me also thank and congratulate @Aaron-Robertson for all the work done in developing most of the DDD module, from the initial RFC up to the final MVP!

**Pull request merged**
